### ISSUE TITLE
[Question] MoE load-balancing monitoring in veRL (Megatron training + SGLang rollout) + where to add load-balance loss?

### ISSUE THREAD

**Issue by tiesanguaixia** [verl-project/verl#5232]
Hi verl team — quick questions about MoE training:

**1. Monitoring**
Does verl have built-in logging for MoE load balancing metrics (e.g., tokens per expert, router entropy, dropped tokens)? If not, where should we add custom monitoring?

**2. Load-balance loss**
I see `moe_router_load_balancing_type="none"` in [config_converter.py](https://github.com/verl-project/verl/blob/main/verl/models/mcore/config_converter.py#L202). Is this the recommended setting for RL training? If we want to enable aux loss for better expert utilization, what's the suggested approach — just set `moe_aux_loss_coeff` via `override_transformer_config`, or modify the training code directly?

Thanks!

**Comment by Kirrito-k423** [comment]
Good questions. Here's the current state:

**1. MoE Monitoring**

veRL doesn't have built-in MoE-specific metrics logging. To add custom monitoring, the best place is in the Megatron forward pass:

- `verl/models/mcore/model_forward.py` - Add hooks to capture routing decisions
- `verl/workers/megatron_workers.py` - Log metrics during training step

You can access the router's expert selection indices and compute:
- Tokens per expert distribution
- Router entropy: `-sum(p * log(p))` where p is routing probability
- Load imbalance coefficient

**2. Load-balance Loss**

The `moe_router_load_balancing_type="none"` is the default because:
- RL training already has a complex optimization landscape
- Auxiliary load-balance loss can interfere with the RL objective

To enable it:
```yaml
actor_rollout_ref.actor.megatron.override_transformer_config:
  moe_router_load_balancing_type: "aux_loss"
  moe_aux_loss_coeff: 0.01  # Start small
```

For fine-grained control, you'd need to modify `verl/models/mcore/model_forward_fused.py` to capture and return the aux loss component separately, then add it to the PPO loss.

Note that enabling aux loss may require adjusting your RL hyperparameters (lower lr, different kl_coef) since it changes the gradient dynamics.

**Comment by Kirrito-k423** [comment]
中文：补充一下当前 `main` 分支的状态：verl 现在已经有一条可选的 rollout 侧 MoE load-balance metrics 路径。相关配置是 `actor_rollout_ref.rollout.enable_rollout_routing_replay` 和 `actor_rollout_ref.rollout.moe_load_balance_metrics_interval`，默认仍关闭；实现会从 `routed_experts` / `response_mask` 统计专家负载，并上报 `rollout/moe/max_vio/*`、`rollout/moe/min_vio/*`、`rollout/moe/avg_vio/*` 等指标。证据：PR #6853、`verl/trainer/ppo/metric_utils.py`、`verl/trainer/config/rollout/rollout.yaml`，以及 `docs/ascend_tutorial/feature_support/ascend_backend_features.md` 中对 `moe_load_balance_metrics_interval` 的说明。

English: Updating this issue with the current `main` branch status: verl now has an opt-in rollout-side MoE load-balance metrics path. It is controlled by `actor_rollout_ref.rollout.enable_rollout_routing_replay` and `actor_rollout_ref.rollout.moe_load_balance_metrics_interval`, and remains disabled by default. The implementation consumes `routed_experts` / `response_mask` and reports metrics such as `rollout/moe/max_vio/*`, `rollout/moe/min_vio/*`, and `rollout/moe/avg_vio/*`. Evidence: PR #6853, `verl/trainer/ppo/metric_utils.py`, `verl/trainer/config/rollout/rollout.yaml`, and the `moe_load_balance_metrics_interval` docs in `docs/ascend_tutorial/feature_support/ascend_backend_features.md`.

中文：关于 load-balance loss，Megatron 的 Qwen2-MoE converter 当前仍把 `moe_router_load_balancing_type` 设为 `"none"`，代码注释说明这是因为 aux loss 会影响 RL 性能。因此它不是默认开启项；如需实验性启用，建议通过 `actor_rollout_ref.actor.megatron.override_transformer_config` 覆盖 router load-balancing 类型和较小的 `moe_aux_loss_coeff`，并同时观察 RL 指标。`examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh` 也展示了 MoE 训练中设置 `moe_aux_loss_coeff=0.01` / `moe_z_loss_coeff=0.001` 的例子。

English: For load-balance loss, the Megatron Qwen2-MoE converter still sets `moe_router_load_balancing_type` to `"none"`, with a code comment saying aux loss is turned off because it hurts RL performance. So it is not enabled by default. If you want to experiment with it, the safer path is to override `actor_rollout_ref.actor.megatron.override_transformer_config`, set the router load-balancing type and a small `moe_aux_loss_coeff`, then monitor RL metrics carefully. `examples/grpo_trainer/run_qwen3_vl_30b_a3b_megatron.sh` also shows an MoE training example using `moe_aux_loss_coeff=0.01` / `moe_z_loss_coeff=0.001`.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
